### DISCUSSION TITLE
Route Types outside of route definotions

### DISCUSSION THREAD

**Discussion by tigawanna** [TanStack/router#656] [Q&A] [answered]
Great project here , am trying to go through the examples to make sense of it all,
was wonderung if theres a way to get the types of the props passed into the component outside f the route definition 

The example with query puts the component in the router function 

```tsx
const postRoute = new Route({
  getParentRoute: () => postsRoute,
  path: '$postId',
  loader: async ({ params: { postId }, context: { queryClient } }) => {
    await queryClient.ensureQueryData(['posts', postId], () =>
      fetchPostById(postId),
    )

    return () =>
      useQuery(['posts', postId], () => fetchPostById(postId), {
        enabled: !!postId,
      })
  },
  component: ({ useLoader }) => {
    const postQuery = useLoader()()

    return (
      <div className="space-y-2">
        <h4 className="text-xl font-bold underline">{postQuery.data?.title}</h4>
        <div className="text-sm">{postQuery.data?.body}</div>
      </div>
    )
  },
})
```

given this route 
```tsx
export const onePostRoute = new Route({
  getParentRoute: () => postsLayout,
  path: "$postId",
  loader: async ({ params: { postId }, context: { queryClient } }) => {
    await queryClient.ensureQueryData({
      queryKey: ["posts", postId],
      queryFn: () => fetchPostById(postId),
    });

    return () =>
      useQuery({
        queryKey: ["posts", postId],
        queryFn: () => fetchPostById(postId),
        enabled: !!postId,
      });
  },
  component:OnePostPage
});
```
Am trying to hoist it out to it's own file but still have the types to the props being passed in , in order to access items like `useLoader` wth the correct typings

```tsx

interface OnePostPageProps extends RouteProps {

}

export function OnePostPage({}:OnePostPageProps){
return (
 <div className='w-full h-full flex items-center justify-center'>

 </div>
);
}

```

i've tried this , but it gives me circular reference errors when i try to use the types , plus it seems to be the entire routes/ type not just the props to be passed into te component

```tsx
type RouteComponent  = typeof onePostRoute
```
Any help?

**Comment by tigawanna** [comment]
also wondering if this is an anti pattern
```tsx
  const loader = onePostRoute.useLoader()()
```

**Comment by pawelblaszczyk5** [comment, marked as answer]
Afaik, Tanner intention is to don't type these props and if you're detaching the component declaration from route declaration you can use different patterns. I think the `onePostRoute.useLoader()()` is the most encouraged and the easiest one. Here is a screenshot of Discord conversation about this (I've decided to post screenshot instead of link in case someone not having an access to Discord):

<img width="1173" alt="image" src="https://github.com/TanStack/router/assets/54995479/6fe40d48-18c1-46b1-8b9f-2bc00c8477d6">

**Reply by tigawanna** [reply]
Nice , guess I'll join the discord
